api.py

```diff
@@ -142,36 +142,113 @@ def index_lecture(self, task_id: str, source: str) -> bool:
         self._update_file_status(file, "Indexed")
         return True
 
+    @post("/index_pdf")
+    def index_pdf(self, file_id: str, source: str) -> bool:
+        file = File.get(self.client, _id=file_id)
+        self._update_file_status(file, "Indexing")
+
+        # For PDFs, we iterate over the blocks (block = page) and then split each chunk of texts into the context
+        # window units.
+
+        documents = []
+
+        for block in file.blocks:
+            # Load the page_id from the block if it exists
+            page_id = None
+            for tag in block.tags:
+                if tag.name == DocTag.PAGE:
+                    page_num = tag.value.get(TagValueKey.NUMBER_VALUE)
+                    if page_num is not None:
+                        page_id = page_num
+
+            for i in range(0, len(block.text), self.config.context_window_size):
+                # Calculate the extent of the window plus the overlap at the edges
+                min_range = max(0, i - self.config.context_window_overlap)
+                max_range = min(len(block.text), i + self.config.context_window_size + self.config.context_window_overlap)
```

Comment:
You could get rid of max_range. 

Python doesn't throw out-of-index errors. Instead, positive, out of bound indices get clipped: 

Pseudocode:
```
a = [1, 2, 3]
a[0:100] == a[0:3]
```